### DISCUSSION TITLE
Unexpected conflict generic extension method signature mapping

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#1295] [General]
I made extension method with signature like this

```C#
public static T ConvertFromPath<T>(this DateTimeOffset date,params string[] paths) where T : class => throw new NotImplementedException();
public static T ConvertFromPath<T>(this DateTimeOffset date,string path,T defaultValue) where T : struct => throw new NotImplementedException();

// then use it like this

var intConvert	= DateTimeOffset.Now.ConvertFromPath("",0);
var string1Convert	= DateTimeOffset.Now.ConvertFromPath<string>("");

// above work fine but below got error

var string2Convert	= DateTimeOffset.Now.ConvertFromPath<string>("","");

// The type 'string' must be a non-nullable value type
// in order to use it as parameter 'T' in the generic type or method
// 'Ext.ConvertFromPath<T>(DateTimeOffset, string, T)'
```

Which I expect that when I put generic type as `string` it should map with the function with `where T : class` which just treat it as `params string[]` but it not. I think this is unexpected behavior of compiler

ps. The actual code is converter from database
ps2. should this be on roslyn?

**Comment by HaloFour** [comment]
Generic constraints aren't currently part of overload resolution.  I believe that this would be covered by the existing championed proposal #98.

**Comment by gafter** [comment]
Constraints cannot be used to disambiguate methods. Yet. As @HaloFour points out, https://github.com/dotnet/csharplang/issues/98 is an open proposal that should make this work.
